### DISCUSSION TITLE
'init(_:content:)' is deprecated: Use the 'IdentifiedArray'-based version, instead.

### DISCUSSION THREAD

**Discussion by inwoodev** [pointfreeco/swift-composable-architecture#1458] [Q&A] [answered]
Hi, there. Currently I am trying to build my Project using TCA and to avoid repetition of `CertificationUploadViews` I tried to use `ForeEachStore` 

The below codes seem to work fine, but I get the warning that says ` 'init(_:content:)' is deprecated: Use the 'IdentifiedArray'-based version, instead.` 

Umm...is there some way to avoid the deprecation warning?

``` swift
// CertificationSectionView.swift
.
.
.
   ForEachStore(self.store.scope(
          state: { $0.certificationUploadStates },
          action: { .certificationUploadAction(index: $0, action: $1)}
        )
        ) { certificationUploadStore in
          CertificationUploadView(store: certificationUploadStore)
          
          CertificationUploadView(store: certificationUploadStore)
        }
```

``` swift
// CertificationSectionCore.swift

struct CertificationSectionState: Equatable {
  var title: String
  var mandatoryNumber: Int
  var certificationUploadStates: [CertificationUploadState]
  var isCertificationEnough: Bool = false
}

enum CertificationSectionAction: Equatable {
  case certificationUploadAction(index: Int, action: CertificationUploadAction)
  case onValidation(index: Int)
}

struct CertificationSectionEnvironment { }
```

**Comment by inwoodev** [comment]
Sorry, should have read documents more carefully. Need id, not index.

**Reply by tgrapperon** [reply]
Great. I kept my response alive as it may serve someone else in the future.

**Comment by tgrapperon** [comment, marked as answer]
Hey @inwoodev. Indeed, `ForEachStore` and `forEach` reducer don't support non-keyed collections like `Array`s for quite some time now. Because indices are not a stable identifier in `Array`, actions could be addressed to the wrong child reducer if one element of the array was removed for example.

To avoid the warning, you should convert your `[CertificationUploadState]` to an `IdentifiedArray` which is more or less like an `OrderedDictionary`, where each element is identified by an identifier instead of an index. If `CertificationUploadState` is `Identifiable`, you can directly create an `IdentifiedArrayOf<CertificationUploadState>` that will use this value's `id` as a key. If it's not the case, and while you can create an `IdentifiedArray` of something that is not `Identifiable`, `Identifiable` values are easier to handle in TCA, and you can wrap your `CertificationUploadState` into an `Identified<ID, CertificationUploadState>` value (this type is also provided by TCA), where `ID` is some unique `Hashable` value you extract from `CertificationUploadState`, like a token for example. If your type has no such value, you can form `Identified<ID, CertificationUploadState>` directly by providing an external `ID` like an `UUID` (you should check the various initializers for `Identified`).

**Reply by inwoodev** [reply]
Thank you fore detailed answer. Indeed, I had to read documents more carefully. With your detailed answer, the problem seems to be more easy to solve :)

**Reply by tgrapperon** [reply]
`Identified` is easy to miss as it's not on TCA's direct path. It can save work in many cases.
